pyswarms/base/base_discrete.py

```diff
@@ -49,7 +49,7 @@ def __init__(
         velocity_clamp=None,
         init_pos=None,
         ftol=-np.inf,
-        ftol_iter=1,
+        ftol_iter=50,
```

Comment:
Also I don't think this is a good change. `ftol_iter`defaults to `1` to preserve backward compatibility with the initial implementation of `ftol`. This is the most basic case of early stopping when we stop as soon as an accepted tolerance is achieved.

Increasing the default value goes against the purpose. I think that the default should be 1 and the user should change the value as per their usecase.

Reply:
This is what I thought before. All reasons for confusions are the naming of ftol_iter.

The current implementation is what I exactly was trying to implement, but with different default values. If you like to keep `ftol_iter=1` and I implement something new, we are using the same algorithm twice.

Reply:
The case that you are talking about is actually pretty common (when er are using random swarm initialization)
> There is one case that delta < ftol may happen at 1st or 2nd iteration, and it's when n_particles is huge, something like 10,000.

 You can test the same by checking  out `tests/optimizers/test_tolerance.py` Just run the `GlobalBestPSO`, make it fail (`assert False`) and update the code to print the delta at each iteration.

I've generated the below sample output the same way (running the current tests) and added a `print(i, delta)` to the `global_best.py` file.

1. `test_ftol_effect (ftol = 0.01)`
```py
0 False
1 False
2 False
3 True    # <- stops immediately
```
2. `test_ftol_iter_effect (ftol = 0.01, ftol_iter=50)`
```py
0 False
1 True    # <- would have stopped here (if ftol_iter=1)
2 True
3 False
4 False
5 True
6 True
7 False
8 True
9 True
10 True
11 True
12 True
...
57 True    # <- finally stops at iteration 57 (after 50 acceptable iterations)
```
Additionally, in the library delta is not defined as the difference between two consecutive iterations. It is the difference between the `current_best_cost` and the `global_best_cost`. So for the `stall_iter` (`n_iter_no_change`) condition, we need to check the improvements in the `global_best_cost` (not delta)

```py
# delta calculation
relative_measure = self.ftol * (1 + np.abs(best_cost_yet_found))
delta = (
    np.abs(self.swarm.best_cost - best_cost_yet_found)
    < relative_measure
)
```

@msat59 Let me know if you have any issues replicating these test results. @ljvmiranda921 Please do point out if I missed anything.

Reply:
@nishnash54, this is what I saw last night in **Tolerance** jupyter notebooks in docs\examples\tutorials folder.

`ftol_iter` helps to generalize the ftol property and extends the feature over 'n' iterations. As a fixed tolerance might seem too restrictive, ftol_iter provides the optimizer some breathing space to improve the best cost. If there are no improvements detected over ftol_iter subsequent iterations, the optimizer stops.

Reply:
I hear what you say @nishnash54. but...

when there is no improvement (stalled), the `best_cost_yet_found` is actually `self.swarm.best_cost`. so delta gets zero, which is less than `relative_measure`.

Here is the results at 4th iteration, in my test case.

There is no improvement in last two iterations.
![image](https://user-images.githubusercontent.com/20813541/108606557-61b4fc00-73c3-11eb-8ce0-a92b2125edce.png)

and the checks are
![image](https://user-images.githubusercontent.com/20813541/108606632-bf494880-73c3-11eb-9be3-2a0d4d694d17.png)
![image](https://user-images.githubusercontent.com/20813541/108606582-87420580-73c3-11eb-8a46-b32e69bfff89.png)

You see... `self.swarm.best_cost = best_cost_yet_found` so **delta is True** because np.abs(0.023167180458767107 - 0.023167180458767107) = 0 < 1.0231671804587672e-06.

so deque([False, False, **True**, **True**], maxlen=50)

Reply:
The numbers you have presented are correct. That is the actual cost history of the swarm over the iterations. But `ftol` does not deal with the actual costs. The cost based early stopping is not yet implemented. `ftol` is a relative measure.

The stopping criterion is defined by the improvement on the `best_cost_yet_found`. So if the change is big enough the iterations continue.

```py
# print(i, self.swarm.best_cost, best_cost_yet_found, delta)
22 -56.0 -56.0 True
23 -56.0 -56.0 True
24 -56.0 -56.0 True
25 -58.0 -56.0 False  # <- Here the change is big enough to let the algorithm continue
26 -58.0 -58.0 True
27 -58.001 -58.0 True  # <- Here there is a small improvement (not big enough for delta to change)
                       # This is the exact case where we still want to consider the iterations for
                       # early_stopping, but on the other hand the algorithm has not stalled.
28 -58.001 -58.001 True
29 -58.001 -58.001 True
```
Again this implementation is early_stopping based and not based on a cutoff threshold (as that will change for each optimization problem). If you're having problems running the tests cases, please try running through the logic with some sample values in a list.

> I think it is better to withdraw this PR and submit one with this implementation to have both stall_iter and ftol_iter. I didn't want to make things complicated but it seems there is no other choice.

@ljvmiranda921 Let us know how you want to proceed with this.

Reply:
> The numbers you have presented are correct. That is the actual cost history of the swarm over the iterations. But ftol does not deal with the actual costs. The cost based early stopping is not yet implemented. ftol is a relative measure.

@nishnash54, I know that ftol deals with differences in costs and relative measure. I took those screenshots to show you that the current implementation deals with the difference between **two consecutive iterations** which counts stalled iterations.

If I implement it in my own way, I have to change your implementation to exclude stalled iterations. I'm sure you complain later that it's not working as you have implemented ftol_iter. 

Guys, I think we have a conflict of interest here. :wink:
I feel you are against any changes in your implementation, which has resulted in hundreds of lines of explanations and then may cause the introduction of complexity in PySwarms functionality.

I'm happy with your implementation, however, it is not 100% efficient computationally. If the iters=10000 and ftol_iter is 10000 for instance, an array of 10k boolean values must be processed 10k times (all(ftol_history), which is not an efficient way; but I didn't complain. I requested for renaming `ftol_iter` to `stall_iter` but you were not happy. Then I kept whatever we had and just changed the default values and then you were not happy.

Please tell me what I should do. I'm really stalled. :laughing: 
If I forget about any changes I made now and withdraw the PR, and implement **fmin** feature, I need a `stall_iter` parameter, which is exactly what you've named it as `ftol_iter`. And it's default value must be greater than 1 to avoid premature optimization, which is what ftol_iter does at the moment.

**Please explain why should we have a duplicated parameter?**

If I implement my own way and keep ftol_iter as it is (there will be some conflicts), it is like having a car, using its boot as a storage container, and wanting to buy a new and exactly the same car because we want to drive to work.

That's the story guys. Please advise.
@nishnash54 
@ljvmiranda921

Reply:
I'm good with the implementation. I just pointed out 2 cases to take into consideration
 1. Changing the default `ftol_iter=50` will break backward compatibility. (This has got nothing to do with any logic)
 2. The `ftol_iter` already handles the stalling part as you have correctly mentioned. But the algorithm is not exactly "stalled" but is not improving enough, so I suggested `n_iter_no_change`.

I'm just going by the understanding from the initial issue where @D4T4-80Y mentioned the use case as
> The stall means no improvement in the best_cost no matter what its value is, which means it **stays constant in two consecutive iterations** . In other word, diff=0 which is always less than ftol.

Again I am fine with the PR as it does not affect the working of `ftol` and `ftol_iter`, but I don't think it solves the issue pointed out in #447 @msat59 Sorry if I wasn't clear about this. :expressionless:

Reply:
I'll get back to these this weekend, sorry for the delay (got busy with work). Let's be on-track re: Option B that we agreed upon in the original Issue. Also, let's do our best to keep the conversation as calm and objective as possible. cc: @nishnash54 @msat59

Reply:
@ljvmiranda921 , we are all good, I think. :wink: